Feature - 5512 - Individual roles user edit

esizer opened the pull request:
🤖 Resolves #5512 

## 👋 Introduction

Adds a table with the users individual roles to the edit user page in admin.

## 🕵️ Details

This contains the table along with two dialogs, one to add a role and another to remove a role.

## 📋 TO DO

- [x] Merge in role mutation/queries
- [x] Update to use new mutations/queries
- [x] Send off for translations
- [x] Add translations

## 🧪 Testing

Assist reviewers with steps they can take to test that the PR does what it says it does.

1. Build the app `npm run build`
2. Navigate to `/admins/users`
3. Click "Edit" on a user
4. Confirm the roles table contains the users existing roles
5. Confirm adding a role works
6. Confirm removing a role works

## 📸 Screenshot

Add a screenshot (if possible).

vd1992 commented:
for some reason, Pool Operator is being attached twice to `admin@test.com`
guessing that is a seeding oopsie introduced somewhere

![image](https://user-images.githubusercontent.com/40485260/222852833-28723433-bcba-4d07-992b-7fe0aa0746b2.png)

vd1992 commented on apps/web/src/pages/Users/UpdateUserPage/components/IndividualRoleTable.tsx:
```diff
@@ -0,0 +1,103 @@
+    <Table
+      data={data}
+      columns={columns}
+      addDialog={
+        <AddIndividualRoleDialog
+          userName={userName}
+          availableRoles={availableRoles}
+          onAddRoles={handleAddRoles}
```
need to filter out team-based roles somewhere, before passing into the individual role dialog

![image](https://user-images.githubusercontent.com/40485260/222853186-041e6f6b-7438-40ab-bf35-9cde8d36eaa5.png)

esizer replied:
Filtered out team based roles in [filter out team based roles and current user roles](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5799/commits/0f56ac40106f72881e701ae9633a5a64a41ccb4e)

esizer commented:
> for some reason, Pool Operator is being attached twice to `admin@test.com`
> guessing that is a seeding oopsie introduced somewhere

Yup, but this is totally valid. However, this is for individual roles so this is a bug I need to resolve since no team roles should appear here.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/c297c6e72193deec70af5c31eecc16eebb33da82/api/database/seeders/UserSeederLocal.php#L48-L49

vd1992 commented on apps/web/src/pages/Users/UpdateUserPage/components/IndividualRoleTable.tsx:
```diff
@@ -0,0 +1,103 @@
+  );
+
+  // TO DO: Update to user roles
+  const data = useMemo(() => {
+    const roles = user.roleAssignments
+      ?.filter(notEmpty)
+      .map((assignment) => assignment.role)
+      .filter(notEmpty);
+    return roles || [];
+  }, [user.roleAssignments]);
```
need to filter out team based roles so pool operator does not appear

vd1992 replied:
I see your comment

> Yup, but this is totally valid. However, this is for individual roles so this is a bug I need to resolve since no team roles should appear here.

vd1992 replied:
oh, and maybe add the headers for the individual role table section?

![image](https://user-images.githubusercontent.com/40485260/222854805-3191f553-c72f-4edf-9072-85975ffd47d8.png)

per

![image](https://user-images.githubusercontent.com/40485260/222854904-0614964c-735e-407c-8d71-7dee037d23a0.png)

esizer replied:
> Should individual roles disappear from the list when they're already added? If not now, should we spin that off into a separate issue? (same for removal of roles)

Removed current user roles in [filter out team based roles and current user roles](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5799/commits/0f56ac40106f72881e701ae9633a5a64a41ccb4e)

esizer replied:
Added headings in [add headings to roles section](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5799/commits/14e93aadab56ba2a56136a4d99460ca27288c18b)

patcon commented:
Should there be some text or even just an empty space when the table is empty?

<img width="1139" alt="Screen Shot 2023-03-06 at 1 01 17 PM" src="https://user-images.githubusercontent.com/305339/223117496-01a8ee47-beb8-489f-ae65-f42f9b79e989.png">

patcon commented:
Should individual roles disappear from the list when they're already added? If not now, should we spin that off into a separate issue? (same for removal of roles)

<img width="1127" alt="Screen Shot 2023-03-06 at 1 05 59 PM" src="https://user-images.githubusercontent.com/305339/223118694-d1b7e017-5ec8-4164-b5d7-03462c646147.png">

patcon commented:
If we do want to have all roles listed in the "Add Role" modal (as opposed to filtering out ones that exist), do we want a validation error when a role is added that was already there? Right now, it just shows as "added" toast whether it was truly added or not:
https://recordit.co/MQE5nQXRGZ

esizer commented:
Should individual roles disappear from the list when they're already added? If not now, should we spin that off into a separate issue? (same for removal of roles)

Removed current user roles in [filter out team based roles and current user roles](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5799/commits/0f56ac40106f72881e701ae9633a5a64a41ccb4e)

patcon commented on apps/web/src/pages/Users/UpdateUserPage/components/AddIndividualRoleDialog.tsx:
```diff
@@ -0,0 +1,167 @@
+
+  const roleOptions = availableRoles
+    .filter((role) => {
+      return !(
+        !role.isTeamBased ||
+        !user?.roleAssignments?.some(
+          (assignment) => assignment.role?.id === role.id,
+        )
+      );
```
We only want roles that are (1) not team-based and (2) aren't in the user's current list of roleAssignments.

```suggestion
      return (
        !role.isTeamBased &&
        !user?.roleAssignments?.some(
          (assignment) => assignment.role?.id === role.id,
        )
      );
```

(This may have appeared correct bc the graphql query was missing isTeamBased on the query response)

patcon commented on apps/web/src/pages/Users/userOperations.graphql:
```diff
@@ -598,3 +610,14 @@ query selectedUsers($ids: [ID]!) {
+
+query ListRoles {
+  roles {
+    id
+    displayName {
+      en
+      fr
+    }
+    name
```
```suggestion
    id
    name
    isTeamBased
    displayName {
      en
      fr
    }
```

patcon commented:
modal styling changed unexpectedly when branch was updated. re-running chromatic job to see if it also shows this diff from mainline

<img width="1406" alt="Screen Shot 2023-03-06 at 5 03 56 PM" src="https://user-images.githubusercontent.com/305339/223180458-183ce687-8281-46a6-8461-d3c89aa091f6.png">

esizer commented:
> modal styling changed unexpectedly when branch was updated. re-running chromatic job to see if it also shows this diff from mainline

Fixed with [add dialog body wrapper](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5799/commits/8f43616d5b77fd03f744fe13e7d1462ad99ec16d)

vd1992 commented:
for funny cases, is stopping someone from removing platform admin from themselves something to have, preventing odd situations
in general, are there any issues with editing your own roles to watch out for?

esizer commented:
> for funny cases, is stopping someone from removing platform admin from themselves something to have, preventing odd situations

Definitely seems odd but I don't think we should prevent it. I could see a use case where someone would want to remove themselves as an admin. Also, that is why we have the confirmation dialog (to help prevent accidental destructive actions) 😄

vd1992 commented:
I'm getting an odd visual bug where the contents of the select dropdown don't all appear if you have a user with all roles available to add. This makes it tricky to get to a role you may want to assign, seems to be due to the dropdown truncating at the dialog's edge. Can't scroll down the select either. 

![gif](http://g.recordit.co/jubHY4TIY4.gif)

esizer commented:
> I'm getting an odd visual bug where the contents of the select dropdown don't all appear if you have a user with all roles available to add. This makes it tricky to get to a role you may want to assign, seems to be due to the dropdown truncating at the dialog's edge. Can't scroll down the select either.

Good catch, looks like at some point we added `overflow: hidden;` to the dialogs. As you can see, this causes issues in certain cases so I have removed it in [remove overflow hidden from dialogs](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5799/commits/e762dae8512dab455e86dd8018cb727bdc66a4eb)

esizer commented:
> Would it be more appropriate for there to be some text or even just an empty space when the table is empty with no roles?

I agree, @patcon ! However, currently this is how all of our admin tables work so it might be better to create an issue to do a proper design and implementation of tables when they are empty 👍

vd1992 reviewed (approved):
not sure if Pat has more to add, and translations are pending
but other than that, I got nothing more to add

great!

patcon reviewed (approved):
Yeah, this looks great! Thanks @esizer! (and good catch on the overflow @vd1992 )

mnigh commented on apps/web/src/lang/fr.json:
```diff
@@ -6646,6 +6646,66 @@
+    "description": "Lead in text for the add role to user form."
+  },
+  "UGk7lG": {
+    "defaultMessage": "Enregistrer les changements",
```
We have an identical string in English that is translated differently. Which should we use?

https://github.com/GCTC-NTGC/gc-digital-talent/blob/7587d2b1a393b7c19d5f172fec3b7208a7f62d93/apps/web/src/pages/PoolCandidates/ViewPoolCandidatePage/components/ApplicationStatusForm/ApplicationStatusForm.tsx#L212-L217

https://github.com/GCTC-NTGC/gc-digital-talent/blob/7587d2b1a393b7c19d5f172fec3b7208a7f62d93/apps/web/src/lang/fr.json#L4641-L4644

mnigh replied:
> Seems like this is a common string we use. Maybe it should be moved to [`packages/i18n/` ](https://github.com/GCTC-NTGC/gc-digital-talent/blob/main/packages/i18n/src/messages/formMessages.ts) 🤔

Yes, that makes sense. I'm not sure which one should be used though. @substrae any thoughts?

esizer replied:
Based on discussion, moved to `packages/i18n` and chose "Enregistrer les modifications" in [extract save changes message into i18n package](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5799/commits/aceaf36882712c223434823c69a6217e814794c6)

esizer commented on apps/web/src/lang/fr.json:
```diff
@@ -6646,6 +6646,66 @@
+    "description": "Label for the form to add a role to a user"
+  },
+  "HYYYtr": {
+    "defaultMessage": "L’utilisateur perdra le rôle suivant :",
```
Do we have anyone who could weigh in on this right now?

esizer replied:
Updated message in [update remove role dialog message](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5799/commits/103fbed779ae9b9bd979881e00262a821b72240a)

mnigh commented on apps/web/src/pages/PoolCandidates/ViewPoolCandidatePage/components/ApplicationStatusForm/ApplicationStatusForm.tsx:
```diff
@@ -209,12 +210,7 @@ export const ApplicationStatusForm = ({
               color="primary"
               isSubmitting={isSubmitting}
-              text={intl.formatMessage({
-                id: "OunUSG",
```
You'll need to remove `OunUSG` from `apps/web/src/lang/fr.json`.

esizer replied:
Removed in [trim unused translations](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5799/commits/0621b4a23be5afc6da9c07cdf7b12a6caf88acd0)

mnigh reviewed (changes requested):
Almost there!
